aezeed/cipherseed.go

```diff
@@ -0,0 +1,484 @@
+	// aez.
+	keyLen = 32
+
+	bitsPerWord = 11
+)
+
+var (
+	scryptN = 32768
+	scryptR = 8
+	scryptP = 1
+)
+
+var (
+	// bitcoinGenesisDate is the timestamp of Bitcoin's genesis block.
+	// We'll use this value in order to create a compact birthday for the
+	// seed. The birthday will be interested as the number of days since
+	// the genesis date. We refer to this time period as ABE (after Bitcoin
+	// era).
+	bitcoinGenesisDate = time.Unix(1231006505, 0)
+)
+
+// CipherSeed is a fully decoded instance of the aezeed scheme. At a high
+// level, the encoded cipherseed is the enciphering of: a version byte, a set
+// of bytes for a timestamp, and finally the entropy which will be used to
+// directly construct the HD seed. This scheme was created as the widely used
+// schemes in the space lack two critical traits: a version byte, and a
+// birthday timestamp. The version allows us to modify the details of the
+// scheme in the future, and the birthday gives wallet a limit of how far back
+// in the chain they'll need to start scanning. We also add an external version
+// to the enciphering plaintext seed. With this addition, seeds are able to be
+// "upgraded" (to diff params, or entirely diff crypt), while maintain the
+// semantics of the plaintext seed.
+//
+// The core of the scheme is the usage of aez to carefully control the size of
+// the final encrypted seed. With the current parameters, this scheme can be
+// encoded using a 24 word mnemonic. We use 8 bytes of ciphertext expansion
+// when enciphering the raw seed, giving us the equivalent of a checksum.
+// Finally, the password in the scheme is optional. If not specified, an empty
+// string will be used as the password. Otherwise, the addition of the password
+// means that users can encrypt the saw "plaintext" seed under distinct
+// passwords to produce unique mnemonic phrases.
+type CipherSeed struct {
+	// InternalVersion is the version of the plaintext cipherseed. This is
+	// to be used by wallets to determine if the seed version is compatible
+	// with the derivation schemes they know.
+	InternalVersion uint8
+
+	// Birthday is the time that the seed was created. This is expressed as
+	// the number of dates since the timestamp in the Bitcoin genesis
+	// block. We use days as seconds gives us wasted granularity. The
+	// oldest seed that we can encode using this format is through the date
+	// 2188.
+	Birthday uint16
+
+	// Entropy is a set of bytes generated via a CSPRNG. This is the value
+	// that should be used to directly generate the HD root, as defined
+	// within BIP0032.
+	Entropy [EntropySize]byte
+
+	// salt is the salt that was used to generate the key from the user's
+	// specified passphrase.
+	salt [saltSize]byte
+
+	// cipherText is the cached cipher text. This is the result of
+	// enciphering the above with aez, applying our versioning, salt, and
+	// associated data authentication scheme.
+	cipherText [EncipheredCipherSeedSize]byte
```

Comment:
This doesn't seem to be used other than setting it when computing the ciphertext. If it's meant to be a cache, then we'd want to check if it was set before trying to compute it.

Reply:
Yeah I had an initial use for it, but changed my mind towards the end. Updated to remove the field all together.